keel-orca/src/test/kotlin/com/netflix/spinnaker/keel/constraints/PipelineConstraintEvaluatorTests.kt

```diff
@@ -0,0 +1,171 @@
+import dev.minutest.junit.JUnit5Minutests
+import dev.minutest.rootContext
+import io.mockk.coEvery
+import io.mockk.mockk
+import io.mockk.slot
+import strikt.api.expectThat
+import strikt.assertions.isEqualTo
+import strikt.assertions.isFalse
+import strikt.assertions.isLessThanOrEqualTo
+import strikt.assertions.isNotNull
+import strikt.assertions.isNull
+import strikt.assertions.isTrue
+import java.time.Clock
+
+internal class PipelineConstraintEvaluatorTests : JUnit5Minutests {
+
+  companion object {
+    val clock: Clock = Clock.systemUTC()
+    val deliveryConfigRepository = InMemoryDeliveryConfigRepository(clock)
+  }
+
+  data class Fixture(
+    val constraint: PipelineConstraint
+  ) {
+    val type = "pipeline"
+    val orcaService: OrcaService = mockk()
+    val artifact = DeliveryArtifact("fnord", ArtifactType.DEB)
+    val version = "1.1"
+    val environment = Environment(
+      name = "prod",
+      constraints = setOf(constraint)
+    )
+    val manifest = DeliveryConfig(
+      name = "my-manifest",
+      application = "fnord",
+      artifacts = setOf(artifact),
+      environments = setOf(environment)
+    )
+    val executionId = randomUID().toString()
+    val capturedId = slot<String>()
+    val trigger = slot<Map<String, Any?>>()
+    val subject = PipelineConstraintEvaluator(orcaService, deliveryConfigRepository, clock)
+  }
+
+  fun tests() = rootContext<Fixture> {
+    fixture {
+      Fixture(
+        constraint = PipelineConstraint(
+          pipelineId = randomUID().toString(),
+          parameters = mapOf("youSayManaged" to "weSayDelivery")
+        ))
+    }
+
+    context("promotion is gated on launching a pipeline and its final status") {
+      before {
+        deliveryConfigRepository.store(manifest)
+
+        coEvery {
+          orcaService.triggerPipeline(any(), capture(capturedId), capture(trigger))
+        } returns TaskRefResponse("/pipelines/$executionId")
+      }
+
+      test("first pass persists state and triggers the pipeline") {
+        expectThat(subject.canPromote(artifact, version, manifest, environment))
+          .isFalse()
+
+        expectThat(capturedId.captured)
+          .isEqualTo(constraint.pipelineId)
+
+        expectThat(trigger.captured)
+          .isEqualTo(
+            mapOf(
+              "youSayManaged" to "weSayDelivery",
+              "type" to "managed",
+              "user" to "keel"))
+
+        val state = deliveryConfigRepository.getConstraintState(manifest.name, environment.name, version, type)
+        expectThat(state).isNotNull()
+
+        val attributes = state!!.attributes as PipelineConstraintStateAttributes
+        with(attributes) {
+          expectThat(executionId).isEqualTo(fixture.executionId)
+          expectThat(attempt).isEqualTo(1)
+          expectThat(latestAttempt).isLessThanOrEqualTo(clock.instant())
+          expectThat(lastExecutionStatus).isNull()
+        }
+      }
```

Comment:
I'd love to get your thoughts on how Strikt could help make this more expressive. You _can_ do it with out the `!!` but it's kinda syntactically obscure.

```kotlin
val state = deliveryConfigRepository.getConstraintState(manifest.name, environment.name, version, type)
expectThat(state)
  .isNotNull() // now the subject is known to be non-null
  .isA<PipelineConstraintStateAttributes>() // now the subject is known to be an instance of this
  .and { 
    get { executionId }.isEqualTo(fixture.executionId)
    get { attempt }.isEqualTo(1)
    get { latestAttempt }.isLessThanOrEqualTo(clock.instant())
    get { lastExecutionStatus }.isNull()
  }
```

Reply:
This is great, thank you! I still have a lot to learn about strikt, will update around this pattern and keep it in mind for the future.